tests/InferTypesFuzz.elm

```diff
@@ -0,0 +1,329 @@
+                |> Shrink.map Canonical.List
+                |> Shrink.map located
+
+        Canonical.If { then_, else_ } ->
+            -- TODO: Shrink the test once we get more interesting Bool expressions.
+            [ then_, else_ ]
+                |> List.map shrinkTo
+                |> concatShrink
+                |> shrink expr
+
+        Canonical.Lambda { argument, body } ->
+            expr
+                |> shrinkLambda argument body
+
+        _ ->
+            expr |> Shrink.noShrink
+
+
+shrinkLiteral : Shrinker Literal
+shrinkLiteral lit =
+    case lit of
+        Int i ->
+            i |> Shrink.int |> Shrink.map Int
+
+        Float f ->
+            f |> Shrink.float |> Shrink.map Float
+
+        Char c ->
+            c |> Shrink.char |> Shrink.map Char
+
+        Bool b ->
+            b |> Shrink.bool |> Shrink.map Bool
+
+        String s ->
+            s |> Shrink.string |> Shrink.map String
+
+
+shrinkPlus : Canonical.LocatedExpr -> Canonical.LocatedExpr -> Shrinker Canonical.LocatedExpr
+shrinkPlus left right _ =
+    lazyMap2 Canonical.Plus
+        (exprShrinker left)
+        (exprShrinker right)
+        |> Shrink.map located
+
+
+shrinkLambda : VarName -> Canonical.LocatedExpr -> Shrinker Canonical.LocatedExpr
+shrinkLambda argument body _ =
+    body
+        |> exprShrinker
+        |> Shrink.map (Canonical.lambda argument)
+        |> Shrink.map located
+
+
+shrinkNonEmptyList : Shrinker a -> Shrinker (List a)
+shrinkNonEmptyList shrinkElement list =
+    case list of
+        first :: rest ->
+            lazyMap2 (::)
+                (shrinkElement first)
+                (Shrink.list shrinkElement rest)
+
+        [] ->
+            Shrink.noShrink list
+
+
+shrinkTo : a -> Shrinker a
+shrinkTo shrunk _ =
+    True
+        |> Shrink.bool
+        |> Shrink.map (always shrunk)
+
+
+{-| Calls a shrinker with the supplied argument.
+
+Sometimes you build up a shrinker based on the value to shrink.
+Then you need to invoke it.
+Without `shrink` that is
+
+    (value
+        |> someFunc
+        |> someOtherFunc
+    )
+        value
+
+With `shrink` it becomes
+
+    value
+        |> someFunc
+        |> someOtherFunc
+        |> shrink value
+
+---
+
+We cannot write a type annotation here.
+The `LazyList a` type used by shrinkers is not exposed outside `elm-explorations/test`.
+
+    shrink : a -> Shrinker a -> Shrinker a
+
+-}
+shrink expr shrinker =
+    shrinker expr
+
+
+{-| Combines two lazy lists using a combining function.
+
+---
+
+We cannot write a type annotation here.
+The `LazyList a` type used by shrinkers is not exposed outside `elm-explorations/test`.
+
+    lazyMap2 : (a -> b -> c) -> LazyList a -> LazyList b -> LazyList c
+
+-}
+lazyMap2 f la lb =
+    Shrink.map f la
+        |> Shrink.andMap lb
+
+
+concatShrink : List (Shrinker a) -> Shrinker a
+concatShrink shrinkers =
+    shrinkers
+        |> List.foldl Shrink.merge Shrink.noShrink
```

Comment:
With these two functions, we can probably start our own `Shrink.Extra` module. Then they can be named `map2` and `concat`. (Also, the `shrinkTo`, which I'd name `constant` if it does what I think it does, can move there too!)